core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -210,7 +214,9 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
     if (producerEpoch == Short.MaxValue)
       throw new IllegalStateException(s"Cannot fence producer with epoch equal to Short.MaxValue since this would overflow")
 
-    prepareTransitionTo(PrepareEpochFence, producerId, (producerEpoch + 1).toShort, RecordBatch.NO_PRODUCER_EPOCH, txnTimeoutMs,
+    val bumpedEpoch = if (hasFailedEpochFence) producerEpoch else (producerEpoch + 1).toShort
```

Comment:
I believe it's accurate to say that if `hasFailedEpochFence` is set, then the bumped epoch could not have been returned to the client. Is that right? It might be worth a comment emphasizing that.

Reply:
Added